### PR TITLE
Do not check parameter modifiers during parsing.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This PR specifically makes it so that modifiers on parameters are not checked (while parsing). Previously, we disallowed (at parse time) certain modifier combinations. There's no reason for this complexity and specificity to be at the parser level. Syntactically we know what's going on, and we can allow this while parsing and check and report the error later. It also prevents incrementally reusing parameters (and their containers) when this happens.

Tagging @dotnet/roslyn-compiler @gafter @VSadov

This part of a long term effort to remove unnecessary errors from the parser and move them to later stages of the compiler pipeline. Doing this makes the parser simpler, makes it faster to get a syntax tree, and makes our incremental implementation simpler (because the parser/incremental-parser needs to track less contextual states).

This is also important as syntax-diagnostics end up causing issues for higher level IDE features.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 1]
```diff
@@ -2741,40 +2849,28 @@ public static class Extensions
...
-    //     public static void Foo(int this) {}
-    Diagnostic(ErrorCode.ERR_TypeExpected, ")"),
-    // (10,36): error CS1001: Identifier expected
-    //     public static void Foo(int this) {}
-    Diagnostic(ErrorCode.ERR_IdentifierExpected, ")"));
+            CreateCompilationWithMscorlibAndSystemCore(test).GetDeclarationDiagnostics().Verify(
+                // (10,32): error CS1100: Method 'Foo' has a parameter modifier 'this' which is not on the first parameter
+                //     public static void Foo(int this) {}
+                Diagnostic(ErrorCode.ERR_BadThisParam, "this").WithArguments("Foo").WithLocation(10, 32),
+                // (7,37): error CS1100: Method 'Foo' has a parameter modifier 'this' which is not on the first parameter
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: >// (7,37): error CS1100: Method 'Foo' has a parameter modifier 'this' which is not on the first parameter [](start = 16, length = 105)

Consider asserting parse errors as well to make sure that error experience for the line 7 isn't degraded.

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs [hunk 2]
```diff
@@ -31,11 +32,11 @@ internal static class ParameterHelpers
...
             {
-                SyntaxToken outKeyword;
-                SyntaxToken refKeyword;
-                SyntaxToken paramsKeyword;
-                SyntaxToken thisKeyword;
-                var refKind = GetModifiers(parameterSyntax.Modifiers, out outKeyword, out refKeyword, out paramsKeyword, out thisKeyword);
+                CheckParameterModifiers(parameterSyntax, diagnostics);
+
+                var refKind = GetModifiers(parameterSyntax.Modifiers,
+                    out var outKeyword, out var refKeyword,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs]: >var [](start = 24, length = 3)

Consider putting types back.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler Can i get another review?  Thanks!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -4177,145 +4177,25 @@ private static bool CanReuseParameter(CSharp.Syntax.ParameterSyntax parameter)
...
-        }
-
-        private static ParamFlags GetParamFlags(SyntaxKind kind, bool allowThisKeyword)
-        {
             switch (kind)
             {
                 case SyntaxKind.ThisKeyword:
-                    // if (this.IsCSharp3Enabled)
-                    return (allowThisKeyword ? ParamFlags.This : ParamFlags.None);
+                    return allowThisKeyword ? true : false;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Could be `return allowThisKeyword;`?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: that would make a lot of sense :)  will roll into another PR.

**Pull request merged**